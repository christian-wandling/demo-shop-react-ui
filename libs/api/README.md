# api

A comprehensive library for consuming the demo-shop api. Supports accessing a product catalog, user management,
shopping cart operations, and order processing.
