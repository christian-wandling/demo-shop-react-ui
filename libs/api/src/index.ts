export * from './lib/openapi';
export * from './lib/ApiContext';
export * from './lib/services/apiService';
export * from './lib/ApiProvider';
export * from './lib/services/apiService';
export * from './lib/middleware/authMiddleware';
