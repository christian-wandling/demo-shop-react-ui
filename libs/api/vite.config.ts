import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';

export default defineConfig(() => ({
  root: __dirname,
  cacheDir: '../../node_modules/.vite/libs/api',
  plugins: [react()],
  // Uncomment this if you are using workers.
  // worker: {
  //  plugins: [ nxViteTsPaths() ],
  // },
  test: {
    watch: false,
    globals: true,
    environment: 'jsdom',
    reporters: ['default'],
    passWithNoTests: true,
    coverage: {
      reportsDirectory: './test-output/vitest/coverage',
      enabled: false,
    },
  },
}));
