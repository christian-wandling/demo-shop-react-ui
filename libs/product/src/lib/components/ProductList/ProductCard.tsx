import { formatCurrency } from '@demo-shop-react-ui/shared';
import { Link } from 'react-router';

export const ProductCard = ({ product }: { product: any }) => {
  return (
    <Link
      to={`/products/${product.id}`}
      className="product-item group"
      key={product.id}
      data-testid={`product-list-item-link${product.id}`}>
      <div className="aspect-h-1 aspect-w-1 w-full overflow-hidden rounded-lg bg-gray-200 xl:aspect-h-8 xl:aspect-w-7">
        <img
          src={product.thumbnail.uri || 'images/placeholder-image.jpg'}
          alt={product.name}
          onError={e => {
            e.currentTarget.src = 'images/placeholder-image.jpg';
          }}
          className="h-full w-full object-cover object-center group-hover:opacity-75"
          data-testid={`product-list-item-image${product.id}`}
        />
      </div>
      <h3 className="mt-4 text-sm text-gray-700" data-testid={`product-list-item-name${product.id}`}>
        {product.name}
      </h3>
      <p className="mt-1 text-lg font-medium text-gray-900" data-testid={`product-list-item-price${product.id}`}>
        {formatCurrency(product.price)}
      </p>
    </Link>
  );
};
