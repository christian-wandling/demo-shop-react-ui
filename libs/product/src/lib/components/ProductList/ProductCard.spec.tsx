import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router';
import ProductCard from './ProductCard';
import { ProductResponse } from '@demo-shop-react-ui/api';

import { vi } from 'vitest';

vi.mock('../utils/formatters', () => ({
  formatCurrency: vi.fn(price => `$${price.toFixed(2)}`),
}));

describe('ProductCard', () => {
  const mockProduct: ProductResponse = {
    categories: ['category'],
    description: 'description',
    id: 1,
    images: [
      {
        name: 'thumbnail',
        uri: 'https://example.com/thumbnail.jpg',
      },
    ],
    name: 'Test Product',
    price: 19.99,
    thumbnail: {
      name: 'thumbnail',
      uri: 'https://example.com/thumbnail.jpg',
    },
  };

  const renderWithRouter = (ui: React.ReactElement) => {
    return render(ui, { wrapper: BrowserRouter });
  };

  it('renders the product card successfully', () => {
    renderWithRouter(<ProductCard product={mockProduct} />);
    expect(screen.getByText('Test Product')).toBeTruthy();
  });

  it('displays the correct product name', () => {
    renderWithRouter(<ProductCard product={mockProduct} />);
    expect(screen.getByText('Test Product')).toBeTruthy();
  });

  it('displays the formatted price correctly', () => {
    renderWithRouter(<ProductCard product={mockProduct} />);
    expect(screen.getByText('$19.99')).toBeTruthy();
  });

  it('renders the product image with correct src', () => {
    renderWithRouter(<ProductCard product={mockProduct} />);
    const image: HTMLImageElement = screen.getByAltText('Test Product');
    expect(image).toBeTruthy();
    expect(image.src).toBe('https://example.com/thumbnail.jpg');
  });

  it('links to the correct product detail page', () => {
    renderWithRouter(<ProductCard product={mockProduct} />);
    const link: HTMLLinkElement = screen.getByRole('link');
    expect(link.href).toContain('/products/1');
  });
});
