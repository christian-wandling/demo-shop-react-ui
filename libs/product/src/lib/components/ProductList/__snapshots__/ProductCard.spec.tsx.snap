// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard > matches the snapshot 1`] = `
<body>
  <div>
    <a
      class="product-item group"
      data-discover="true"
      href="/products/1"
    >
      <div
        class="aspect-h-1 aspect-w-1 w-full overflow-hidden rounded-lg bg-gray-200 xl:aspect-h-8 xl:aspect-w-7"
      >
        <img
          alt="Test Product"
          class="h-full w-full object-cover object-center group-hover:opacity-75"
          src="https://example.com/thumbnail.jpg"
        />
      </div>
      <h3
        class="mt-4 text-sm text-gray-700"
      >
        Test Product
      </h3>
      <p
        class="mt-1 text-lg font-medium text-gray-900"
      >
        $19.99
      </p>
    </a>
  </div>
</body>
`;
